### ISSUE TITLE
Plack::Middleware::AccessLog::Timed cannot record %D and %T fields properly

### ISSUE THREAD

**Issue by astj** [plack/Plack#549]
# tl;dr

- `Plack::Middleware::AccessLog::Timed` cannot record `%D` and `%T` fields properly
- This is because both `Plack::Middleware::AccessLog::Timed` and `Apache::LogFormat::Compiler` has a bug
- Both of those bugs will make breaking changes

# Background

According to documentation of `Plack::Middleware::AccessLog::Timed`, we can record *the time taken to serve the request* to logfile with `fomat` including `%D` and/or `%T`.
This log is formatted by [`Apache::LogFormat::Compiler`](https://github.com/kazeburo/Apache-LogFormat-Compiler).  In [its document](https://github.com/kazeburo/Apache-LogFormat-Compiler#method), `%T` and `%D` are described as below:
```
%T    custom field for handling times in subclasses
%D    custom field for handling sub-second times in subclasses
```
And according to [Apache's LogFormat templates](http://httpd.apache.org/docs/2.0/mod/mod_log_config.html) (its superset) those format string are described as below:

| Format String | Description |
|----|----|
| `%T` | *The time taken to serve the request, in seconds.* |
| `%D` | *The time taken to serve the request, in microseconds.* |

That means:
- if a request took 2 seconds,
  - `%T` should be `2`
  - `%D` should be `2000000` (`2_000_000`).

# Current behavior

I wrote a test code to check current behavior of `Plack::Middleware::AccessLog::Timed` : https://github.com/astj/Plack/commit/d03ab21bd9b56959d93bb68e708d63ebe5c36630

When I run that test, the test fails ([result](https://gist.github.com/astj/bb8eac5cb13b61d7585e)), and from its result, current behavior looks like follows:

- if a request took 2 seconds,
  - `%T` should be `2`, but currently `2000000`
  - `%D` should be `2000000` (`2_000_000`), but currently `2`

# Investigation

This is caused by a bug in `Plack::Middleware::AccessLog::Timed` and another in `Apache::LogFormat::Compiler`.

## `Plack::Middleware::AccessLog::Timed`

From [documentation of `Apache::LogFormat::Compiler`](https://github.com/kazeburo/Apache-LogFormat-Compiler/blob/291210e72318f188d16f5ccdd2b7ade782640d16/lib/Apache/LogFormat/Compiler.pm#L233-L242), `Apache::LogFormat::Compiler::log_line` expects its argument `$reqtime` to be in microseconds.
But `Plack::Middleware::AccessLog::Timed` passes `$now - $time` in seconds as `$reqtime` argument.
- [`P::M::AccessLog::Timed` passes differences of two `Time::HiRes::gettimeofday`](https://github.com/plack/Plack/blob/2419ca24fc5a8759936c5d17bde3800fde602c10/lib/Plack/Middleware/AccessLog/Timed.pm#L51-L52)
- [`Time::HiRes::gettimeofday` returns *floating seconds* when called in scalar context](https://metacpan.org/pod/Time::HiRes#gettimeofday)

With this commit this bug will be fixed, but due to `Apache::LogFormat::Compiler`'s bug, that test still fails. : https://github.com/astj/Plack/commit/dac66093f454cc0da52ee3333b265f0383f196d7

## `Apache::LogFormat::Compiler`

The bug in `Apache::LogFormat::Compiler` is simple.
When `$reqtime` is passed to `log_line`, `%T` shoule be formatted to `int($reqtime/1_000_000)`, but [with current implementation](https://github.com/kazeburo/Apache-LogFormat-Compiler/blob/291210e72318f188d16f5ccdd2b7ade782640d16/lib/Apache/LogFormat/Compiler.pm#L87-L88) it's formatted to `int($reqtime*1_000_000)`.

I wrote a test code to check this behavior: https://github.com/astj/Apache-LogFormat-Compiler/commit/ea41df7282c0020ad957ed0d72c5190b20a88eec
Unless this bug fixed, `%T` fields in `Plack::Middleware::AccessLog::Timed`'s output remains still broken.

# Suggestion

My suggestion is follows:
- At first, submit a bug-fix pull-request to `Apache::LogFormat::Compiler`
- After that bug fixed and released, fix `Plack::Middleware::AccessLog::Timed` and add dependency of new version of `Apache::LogFormat::Compiler`

**Comment by miyagawa** [comment]
Sounds about right. We should address this. @kazeburo

**Comment by kazeburo** [comment]
Thank you @astj.
I'll fix this bug as soon as possible.

**Comment by miyagawa** [comment]
Fixed by #551
